sw/device/tests/aon_timer_wakeup_irq_test.c

```diff
@@ -0,0 +1,186 @@
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/dif/dif_rv_timer.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/aon_timer_testutils.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/rand_testutils.h"
+#include "sw/device/lib/testing/rv_plic_testutils.h"
+#include "sw/device/lib/testing/test_framework/ottf.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+const test_config_t kTestConfig;
+
+// TODO: [arch,aon] Create a set of target specific AON clock frequency
+// constants #9927
+const uint64_t kClockFreqAonHz = 125 * 1000;  // 125kHz
+
+static const uint32_t kPlicTarget = kTopEarlgreyPlicTargetIbex0;
+static const uint32_t kTickFreqHz = 1000 * 1000;  // 1Mhz / 1us
+static dif_aon_timer_t aon_timer;
+static dif_rv_timer_t rv_timer;
+static dif_rv_plic_t plic;
+
+static volatile dif_aon_timer_irq_t irq;
+static volatile top_earlgrey_plic_peripheral_t peripheral;
+static volatile uint64_t time_elapsed;
+
+/**
+ * Initialize the rv timer to count the tick.
+ *
+ * The `ibex_mcycle_read` function uses the `mcycle` register to count
+ * instructions cycles executed and measure time elapsed between two instants,
+ * however it stops counting when the `wfi` is called. As this test is based on
+ * the `wfi` instruction the best approach then to measure the time elapsed is
+ * to use the mtime register, which is basically attached to rv_timer in the
+ * opentitan.
+ * https://docs.opentitan.org/hw/ip/rv_timer/doc/
+ *
+ * This is fine due to the test running in a single thread of execution,
+ * however, the care should be taken in case it changes. OTTF configures the
+ * timer in vPortSetupTimerInterrupt, and re-initialising it inside the test
+ * could potentially break or cause unexpected behaviour of the test framework.
+ *
+ */
+static void tick_init(void) {
+  CHECK_DIF_OK(dif_rv_timer_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_TIMER_BASE_ADDR), &rv_timer));
+
+  CHECK_DIF_OK(dif_rv_timer_reset(&rv_timer));
+
+  // Compute and set tick parameters (i.e., step, prescale, etc.).
+  dif_rv_timer_tick_params_t tick_params;
+  CHECK_DIF_OK(dif_rv_timer_approximate_tick_params(kClockFreqPeripheralHz,
+                                                    kTickFreqHz, &tick_params));
+
+  CHECK_DIF_OK(
+      dif_rv_timer_set_tick_params(&rv_timer, kPlicTarget, tick_params));
+
+  CHECK_DIF_OK(dif_rv_timer_counter_set_enabled(&rv_timer, kPlicTarget,
+                                                kDifToggleEnabled));
+}
+
+/**
+ * Read the current rv timer count/tick.
+ *
+ * @return The current rv timer count.
+ */
+static uint64_t tick_count_get(void) {
+  uint64_t tick = 0;
+  dif_rv_timer_counter_read(&rv_timer, kPlicTarget, &tick);
+  return tick;
+}
```

Comment:
Might be worth making this a testutil instead. @timothytrippel, wdyt?

Reply:
I totally agree that the test framework should provide an API to measure time, mainly because the implementation may change when the preemption is enabled.

Reply:
Agreed - lets to this in future PR.

Reply:
I've created #9984 to tackle that.